From 66bd3a4c5fd7b0782d5b1ab4b0a8891fe5361764 Mon Sep 17 00:00:00 2001
From: Jason Chen <b02280@freescale.com>
Date: Thu, 16 Dec 2010 11:08:30 +0800
Subject: [PATCH] ENGR00137058 ipuv3:update DP SRM by MCU if FG disable for FG position setting

During FG position set, we used to update SRM by FSU all the time,
if FG is disabled, then after you set position then enable  FG,
it will come out NFB4ENF error for FG channel.

Signed-off-by: Jason Chen <b02280@freescale.com>
---
 drivers/mxc/ipu3/ipu_disp.c |   25 +++++++++++++++++++------
 1 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/drivers/mxc/ipu3/ipu_disp.c b/drivers/mxc/ipu3/ipu_disp.c
index 56cd3d8..17dc616 100644
--- a/drivers/mxc/ipu3/ipu_disp.c
+++ b/drivers/mxc/ipu3/ipu_disp.c
@@ -1802,14 +1802,18 @@ int32_t ipu_disp_set_window_pos(ipu_channel_t channel, int16_t x_pos,
 	u32 reg;
 	unsigned long lock_flags;
 	uint32_t flow = 0;
+	uint32_t dp_srm_shift;
 
-	if (channel == MEM_FG_SYNC)
+	if (channel == MEM_FG_SYNC) {
 		flow = DP_SYNC;
-	else if (channel == MEM_FG_ASYNC0)
+		dp_srm_shift = 3;
+	} else if (channel == MEM_FG_ASYNC0) {
 		flow = DP_ASYNC0;
-	else if (channel == MEM_FG_ASYNC1)
+		dp_srm_shift = 5;
+	} else if (channel == MEM_FG_ASYNC1) {
 		flow = DP_ASYNC1;
-	else
+		dp_srm_shift = 7;
+	} else
 		return -EINVAL;
 
 	if (!g_ipu_clk_enabled)
@@ -1819,8 +1823,17 @@ int32_t ipu_disp_set_window_pos(ipu_channel_t channel, int16_t x_pos,
 
 	__raw_writel((x_pos << 16) | y_pos, DP_FG_POS(flow));
 
-	reg = __raw_readl(IPU_SRM_PRI2) | 0x8;
-	__raw_writel(reg, IPU_SRM_PRI2);
+	if (ipu_is_channel_busy(channel)) {
+		/* controled by FSU if channel enabled */
+		reg = __raw_readl(IPU_SRM_PRI2) & (~(0x3 << dp_srm_shift));
+		reg |= (0x1 << dp_srm_shift);
+		__raw_writel(reg, IPU_SRM_PRI2);
+	} else {
+		/* controled by MCU if channel disabled */
+		reg = __raw_readl(IPU_SRM_PRI2) & (~(0x3 << dp_srm_shift));
+		reg |= (0x3 << dp_srm_shift);
+		__raw_writel(reg, IPU_SRM_PRI2);
+	}
 
 	spin_unlock_irqrestore(&ipu_lock, lock_flags);
 	if (!g_ipu_clk_enabled)
-- 
1.5.4.4

